### PR TITLE
Add Codex Migrate command protection extension

### LINKED ISSUE #2878: [extension] Codex Migrate command protection
### Contribution type

New Extension

### Proposed Extension ID

`command.codex-migrate`

### Capability boundary

Protect the Codex Migrate CLI's destination-changing migration operations. The extension owns the `codex-migrate` executable and reviews the two shell surfaces that can actually change destination state: `export --apply` and `serve --apply`. The latter enables staged transfer, finalization, and verified recovery controls in the local browser. Inventory, inspection, recovery inspection, read-only dashboards, version, and help remain non-reviewable. It does not attempt to inspect HTTP actions taken later inside the local browser UI, nor does it weaken filesystem, SSH, or other matching Guard rules.

### Command surface

Executable: `codex-migrate` (Mac beta/CLI 1.x). Supported subcommands: `launch`, `inventory`, `inspect`, `export`, `serve`, and `recovery`. `export --apply` installs selected repair components; `serve --apply` enables transfer, finalization, and verified recovery controls. `inspect` is read-only even if `--apply` is supplied, while `recovery --apply` is rejected by the CLI. `launch` starts the local guided browser UI without itself applying a transfer.

### Destructive or sensitive examples

- `codex-migrate export --target user@new-mac.local --target-home /Users/user --component personal-skills --apply`
- `codex-migrate serve --target user@new-mac.local --target-home /Users/user --apply`

These forms can install selected components or enable staged migration and recovery changes after the tool's own verification gates.

### Safe counterparts

- `codex-migrate inventory --json`
- `codex-migrate inspect --target user@new-mac.local --target-home /Users/user --json`
- `codex-migrate inspect --target user@new-mac.local --target-home /Users/user --apply` (still read-only)
- `codex-migrate export --target user@new-mac.local --target-home /Users/user --component personal-skills --json`
- `codex-migrate serve --target user@new-mac.local --target-home /Users/user --no-open`
- `codex-migrate recovery --target user@new-mac.local --target-home /Users/user --json`
- `codex-migrate --help`, `codex-migrate --version`, and subcommand help

### Parser and composition edge cases

Tests cover reordered and abbreviated `--apply`, quoted target/home values, paths with spaces, compound commands, `exec` and `xargs` wrappers, help overrides, trailing commands, unresolved expansions in flag position, expansions consumed as known option values, and quoted examples. Unknown or expansion-provided flag arguments must not imply safety.

### Risk and authority model

Proposed action class: `Codex Migrate destination-changing operation`. Risk classes: `destructive_shell`, `execution`, and `network_egress`; severity high; default mode review. Safer alternative: run the same subcommand without `--apply`, inspect the plan and verified destination backup, then repeat with `--apply` only when ready. The extension emits evidence only and remains external/opt-in, so required core and overlapping rules retain their floors.

### Privacy and performance

The matcher is bounded to the canonical executable and explicit subcommand/flag structure. Evidence stores stable IDs and generic details, not targets, usernames, paths, SSH identity files, workspace names, or command text. Read-only invocations are the benign hot path and should exit after bounded token inspection.

### Authoritative references

- https://github.com/jsegeren/codex-migrate
- https://migrate.segeren.com/how-it-works

### Readiness

- [x] I searched the Extension directory and existing issues for overlapping coverage.
- [x] I removed secrets, credentials, private command history, and local paths from these examples.
- [x] I can contribute or help validate destructive and safe-counterpart tests.

### PR DESCRIPTION
**Pull request by jsegeren**
## Summary

- add an external, opt-in `command.codex-migrate` extension
- review `export --apply` and `serve --apply`, the CLI surfaces that can install, transfer, finalize, or restore destination state
- keep inventory, inspection, recovery inspection, read-only dashboards, help, version, and quoted examples quiet
- cover every argparse-accepted long-option abbreviation, `exec`/`xargs` wrapper forms, compound commands, help overrides, and shell-expansion words—including concatenated expansions that can resolve exactly to `--apply`—without recording targets, usernames, paths, or command text
- prove that enabling the extension makes its rule the controlling guard decision, while its default opt-in state remains inert
- publish the regenerated extension-directory catalog and stage the new contribution in packaged artifacts
- include the pre-existing `command.remote.essh` contribution that the staging manifest had omitted

Closes #2878

## Verification

- 512 command-extension, contribution, trust, registry, directory, catalog-identity, and catalog-freshness tests pass
- direct parity check against the real Codex Migrate 1.0.0 argparse surface passes for accepted `export`/`serve` apply aliases and help overrides
- built-wheel inspection confirms the runtime module and contribution manifest are packaged
- `scripts/render_command_extension_directory.py --check`
- `scripts/export_extension_directory.py --check`
- `ruff check`
- `ruff format --check`
- `git diff --check`

A second adversarial review after the automated feedback found and closed an additional concatenated-expansion bypass (`$A$B`, `${DASHES}apply`, `--$MODE`, and command-substitution-plus-literal forms). Regression coverage now includes those cases and near-miss literals that must stay quiet.

I also started the full repository suite. Before I stopped the long run, it reported 726 passes and one reproducible failure in `tests/test_codex_daemon_hook_bridge.py::test_bridge_real_daemon_prefers_payload_cwd_for_verified_git_fetch`: the local live daemon denied `git fetch origin main`. The new external extension is disabled by default and matches only the `codex-migrate` executable, so that daemon/environment failure is unrelated to this contribution.

## Summary by CodeRabbit

- **New Features**
  - Added protection for Codex Migrate destination-changing commands, including `--apply` and abbreviated forms.
  - Detects direct, wrapped, and unresolved shell-expansion invocations.
  - Added safe alternatives and guidance for flagged operations.
  - Registered Codex Migrate as an externally opted-in extension with documentation and packaged metadata.

- **Risk Classification**
  - Destination-changing operations are classified as destructive, executable, and network-egressing.

- **Tests**
  - Added coverage for protected, safe, quoted, wrapped, and expansion-based command scenarios.

### CODE REVIEW

File: src/codex_plugin_scanner/guard/runtime/command_codex_migrate_extensions.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        "--port",
+    }
+)
+_KNOWN_FLAGS = frozenset({"--no-compress", "--apply", "--json", "--no-open", "--help", "-h"})
+_LAUNCHERS: tuple[tuple[str, ...], ...] = (
+    ("codex-migrate",),
+    ("exec", "codex-migrate"),
+    ("xargs", "codex-migrate"),
+)
+_WRAPPER_OPTIONS_WITH_VALUES = frozenset({"-n", "-P", "-I", "-L", "-s"})
```
[the remark this answers is not in the document]
Comment (jsegeren) [reply src/codex_plugin_scanner/guard/runtime/command_codex_migrate_extensions.py]: Fixed in dcd2adff7. The matcher now models separate `exec` and `xargs` option grammars, including value-taking long aliases, attached short values, and wrapper flags. Regression cases cover `exec -a` / `-aNAME`, `xargs --max-args` / `--max-args=`, and `-n1` with expansion-provided apply authority.

File: src/codex_plugin_scanner/guard/runtime/command_codex_migrate_extensions.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+                        argument_index += 1 if separator else 2
+                        continue
+                    if any(marker in argument for marker in _EXPANSION_MARKERS):
+                        evidence.append(
+                            MatcherEvidence(
+                                segment_index=index,
+                                executable=segment.executable,
+                                detail="Matched a Codex Migrate flag-position expansion that may enable apply mode.",
+                            )
+                        )
```
[the remark this answers is not in the document]
Comment (jsegeren) [reply src/codex_plugin_scanner/guard/runtime/command_codex_migrate_extensions.py]: Fixed in dcd2adff7. Expansion review now requires a complete expansion token (including multi-token command substitutions); fixed prefixes or suffixes such as `backup-$HOME` and `${PORT_SUFFIX}-backup` remain non-reviewable. Regression coverage added.

File: src/codex_plugin_scanner/guard/runtime/command_codex_migrate_extensions.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+                argument_index = 0
+                while argument_index < len(remaining):
+                    argument = remaining[argument_index]
+                    if argument == "--":
+                        break
+                    option_name, separator, _value = argument.partition("=")
+                    if option_name in _OPTIONS_WITH_VALUES:
+                        argument_index += 1 if separator else 2
+                        continue
+                    if any(marker in argument for marker in _EXPANSION_MARKERS):
```
[the remark this answers is not in the document]
Comment (jsegeren) [reply src/codex_plugin_scanner/guard/runtime/command_codex_migrate_extensions.py]: Fixed in dcd2adff7. The extension now derives every argparse-accepted unambiguous long-option form per subcommand, uses those forms when consuming option values, and treats accepted help prefixes as safe overrides. Regression cases cover abbreviated value options and `--hel`.

File: contributions/extensions/command.codex-migrate.json [hunk 4]
```diff
@@ -0,0 +1,33 @@
+{
+  "schemaVersion": "guard.extension-contribution.v1",
+  "id": "command.codex-migrate",
```
[the remark this answers is not in the document]
Comment (jsegeren) [reply contributions/extensions/command.codex-migrate.json:3]: Fixed in dcd2adff7. `docs/guard/extensions/catalog.v1.json` is regenerated and committed; `scripts/export_extension_directory.py --check` and the freshness test now pass.

File: src/codex_plugin_scanner/guard/runtime/command_codex_migrate_extensions.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+        action_classes=("Codex Migrate destination-changing operation",),
+        safer_alternatives=(
+            "Run the same subcommand without --apply and inspect the read-only plan first.",
+            "Confirm the destination backup and selected workspace scope before enabling --apply.",
+        ),
+        matcher=_CODEX_MIGRATE_APPLY_WITH_EXPANSIONS,
+        default_mode="review",
+        safe_variants=(
+            safe_flag_variant(
+                _CODEX_MIGRATE_APPLY,
```
[the remark this answers is not in the document]
Comment (jsegeren) [reply src/codex_plugin_scanner/guard/runtime/command_codex_migrate_extensions.py]: Addressed by the current matcher flow in dcd2adff7: `CodexMigrateUnresolvedFlagExpansionMatcher` now checks every accepted help form and exits before emitting expansion evidence. `codex-migrate export --help $FLAGS` and the abbreviated `--hel --apply` case are both covered as safe. The literal matcher keeps its generated help variants because `safe_flag_variant` only accepts executable matcher children.

**Comment by jsegeren** [comment]
@kantorcodes The draft is ready for your review. I verified the Codex Migrate 1.0.0 command surface directly: `export --apply` and `serve --apply` are reviewed; inventory, inspection, recovery inspection, help, and planning remain quiet. The 80-test focused gate and both generated-directory freshness checks pass. I also addressed all four initial automated findings in dcd2adff7.

**Comment by jsegeren** [comment]
Second adversarial pass is now pushed in `0967b5f07`.

It found one bypass beyond the automated review: shell words assembled from multiple expansions or from an expansion plus a literal could resolve exactly to `--apply` without being reviewed. The matcher now handles those forms without broadening to impossible prefixed/suffixed values, and the enabled-control-layer test confirms this opt-in extension actually becomes the controlling rule.

Current evidence: 512 broader command-extension/catalog tests pass, argparse surface parity passes, wheel packaging passes, and catalog/format/lint/diff checks pass. The only observed full-suite failure is the unrelated live-daemon environment denial documented in the PR body.

**Review by kantorcodes — DISMISSED** [review]
The adversarial matcher findings on Codex Migrate appear resolved on this head, including wrapper option grammars, expansion-token precision, argparse abbreviations, help overrides, packaging, and canonical catalog freshness. I do not see a remaining matcher-level blocker in the active threads.

The draft is still missing the managed-control identity regeneration: the changed-file set updates the runtime catalog baseline but not `contracts/managed-controls/v1/extension-projection-digest-vector.json` and its paired delivery constants/tests. Adding the two rules changes that projection identity. Please rebase onto current `main`, regenerate that full digest chain from the final tree, rerun the policy-bundle/extension checks, and then mark ready. That is the remaining blocker I see.

**Comment by jsegeren** [comment]
@kantorcodes Rebased onto current upstream main at dfc2b844d and regenerated the managed-control catalog/projection identity chain from the final tree.

Validation completed locally:
- policy-bundle delivery runtime: 8 passed
- focused Codex Migrate plus extension registry/trust/directory tests: 149 passed
- broader managed-controls, policy-bundle, and catalog regression suite: 426 passed
- generated extension directory check: passed
- Ruff check and format check: passed
- git diff check: passed

The PR is ready for re-review.

**Review by kantorcodes — DISMISSED** [review]
I re-reviewed the new head after the projection/delivery regeneration. That prior blocker is resolved, but this head still has several concrete matcher/privacy blockers before it is safe to approve:

1. Wrapped portable aliases are incomplete: direct matching recognizes `codex-migrate.exe` / `.cmd`, but nested `exec` / `xargs` launcher paths only accept literal `codex-migrate`, leaving Windows wrapper forms uncovered.
2. The argparse prefix model mishandles exact options that are prefixes of longer options. In particular, exact `--target` must still consume its value even though `--target-home` also exists; otherwise a benign expansion in the target value can be treated as an apply-capable flag.
3. `xargs --process-slot-var NAME ...` is a valid value-taking wrapper option and is currently not consumed, so the matcher can lose the inner `codex-migrate` invocation.
4. Expansion feasibility is checked only against the full `--apply` spelling. Argparse accepts abbreviated apply forms too, so concatenated expansions capable of producing `--a` / `--ap` / `--app` need the same fail-secure treatment.
5. Expansion evidence currently copies `segment.executable`; path-qualified launchers can therefore put a local path/username into evidence. Emit the stable executable identifier instead.

Please fix those active review threads and add focused regressions. I am treating the Points Portal managed-control publication note as release coordination, not as a contributor code blocker. After the fixes, rerun the focused Extension/registry/trust/package/policy checks and request re-review.

**Review by kantorcodes — DISMISSED** [review]
The five blockers from my prior review are fixed. One issue remains: the help safe variant still wraps `_CODEX_MIGRATE_APPLY` instead of `_CODEX_MIGRATE_APPLY_WITH_EXPANSIONS`, so expansion-based apply evidence is not covered by the help override. Use the expansion-aware matcher there and add a regression.

File: src/codex_plugin_scanner/guard/runtime/command_codex_migrate_extensions.py [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+        cursor = index + 2
+        while cursor < len(candidate):
+            if candidate.startswith("$(", cursor):
+                depth += 1
+                cursor += 2
+                continue
+            if candidate[cursor] == ")":
+                depth -= 1
+                if depth == 0:
+                    return cursor + 1
```
[the remark this answers is not in the document]
Comment (jsegeren) [reply src/codex_plugin_scanner/guard/runtime/command_codex_migrate_extensions.py:339]: Fixed in 3fc9c41a7. Command-substitution scanning is now quote-, escape-, nested-substitution-, and parenthesis-aware. Regression coverage includes single-quoted and escaped/double-quoted parentheses whose substitution can emit an apply flag.

File: src/codex_plugin_scanner/guard/runtime/command_codex_migrate_extensions.py [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+    fragments.append("".join(literal))
+    return tuple(fragments)
+
+
+def _expansion_end(candidate: str, index: int) -> int | None:
+    if candidate.startswith("${", index):
+        closing = candidate.find("}", index + 2)
+        return closing + 1 if closing >= 0 else None
+    if candidate.startswith("$(", index):
+        depth = 1
```
[the remark this answers is not in the document]
Comment (jsegeren) [reply src/codex_plugin_scanner/guard/runtime/command_codex_migrate_extensions.py]: Fixed in 3fc9c41a7 and hardened in 1c8dc981a. Flag-position words now receive bounded comma-list and Bash-style sequence expansion; any alternative capable of producing an accepted apply form is reviewed, and an over-limit expansion fails secure. The sequence iterator is capped before allocation. Regressions cover --appl{y,}, --appl{a..z}, safe nonmatching alternatives, and a billion-element range.

File: src/codex_plugin_scanner/guard/runtime/command_codex_migrate_extensions.py [hunk 8]
```diff
@@ -0,0 +1,494 @@
...
+    direction = 1 if end_value >= start_value else -1
+    step = direction if step_text is None else int(step_text) if step_text.lstrip("-").isdigit() else 0
+    if step == 0:
+        return None
+    step = abs(step) * direction
+    stop = end_value + direction
+    values = tuple(render(value) for value in range(start_value, stop, step))
+    if not values:
+        return None
+    return values[: _MAX_BRACE_EXPANSION_RESULTS + 1]
```
[the remark this answers is not in the document]
Comment (jsegeren) [reply src/codex_plugin_scanner/guard/runtime/command_codex_migrate_extensions.py]: Fixed in 1c8dc981a. Sequence alternatives are now collected with an early break after the 64-result security bound, before the range can be materialized. A 1..999999999 regression completes through the normal command suite; 1,274 command-security tests pass.

File: contracts/managed-controls/v1/extension-projection-digest-vector.json [hunk 9]
```diff
@@ -1,7 +1,7 @@
...
-  "expectedExtensionProjectionDigest": "sha256:ecaa00e850b845cf89041224ad1fe424b93a096d5b6603373e7580722bf83fe9"
+  "catalogDigest": "d35e6fa975326ad2e8a459a958dfeda1189df2f0269fcdda959955a882d3c90e",
+  "canonicalProjectionJson": "{\"authorityMode\":\"managed-restrictive\",\"delegatedTargets\":[],\"ruleTargets\":[{\"extensionIds\":[\"command.git\"],\"permissionIds\":[\"command.git.permission.force-push\"],\"ruleId\":\"force-push-review\"}],\"schemaVersion\":\"guard.signed-cloud-extension-projection.v1\",\"signedCloudLayersJson\":\"[{\\\"catalog_digest\\\":\\\"d35e6fa975326ad2e8a459a958dfeda1189df2f0269fcdda959955a882d3c90e\\\",\\\"controls\\\":[{\\\"state\\\":\\\"disabled\\\",\\\"target_id\\\":\\\"command.git.permission.force-push\\\",\\\"target_kind\\\":\\\"permission\\\"}],\\\"global_lockdown\\\":false,\\\"kind\\\":\\\"signed-cloud\\\",\\\"schema_version\\\":\\\"1.0.0\\\"}]\"}",
+  "expectedExtensionProjectionDigest": "sha256:e67ba310579e2ed171998a1cdcde5ed72072d92a8dba552a535866573f5a21c0"
```
[the remark this answers is not in the document]
Comment (jsegeren) [reply contracts/managed-controls/v1/extension-projection-digest-vector.json:6]: Acknowledged as release coordination. As Michael noted in the maintainer re-review, this is not a contributor code blocker: the matching Points Portal publication must be staged by maintainers before releasing a runtime that advertises the new digest, while retaining older catalog publications for mixed client versions. This contributor PR intentionally does not mutate the separate Points Portal repository.

**Comment by jsegeren** [comment]
@kantorcodes Re-review requested on dda41310c.

All active matcher and privacy blockers are addressed:
- portable wrapper aliases, exact argparse option precedence, xargs process-slot-var, abbreviated apply expansions, and stable executable evidence
- quote-, escape-, and nested-parenthesis-aware command substitutions
- bounded comma-list and sequence brace expansion, with iteration capped before allocation
- nested parameter expansions such as ${A:-${B:---apply}}
- expansion-aware help safe variants, preserving the accepted help labels
- unrelated remote.essh release staging removed from this PR

Final validation on the pushed tree:
- 1,276 command-security tests passed
- 178 extension, registry, trust, catalog, managed-control, and policy tests passed
- 98 focused Codex Migrate, staging, trust, and directory tests passed
- Ruff lint and format checks passed
- extension-directory freshness and command decision-diff checks passed
- git diff --check passed

The Points Portal publication item remains documented as maintainer release coordination, consistent with your prior review; this contributor PR does not mutate that separate repository.

**Review by kantorcodes — DISMISSED** [review]
The help override still matches `--help` anywhere in the segment. `codex-migrate export ... --apply -- --help` keeps apply active because that `--help` is positional, but the current matcher marks it safe. Please only treat help as an option before `--` and add a regression for this case.

**Comment by jsegeren** [comment]
@kantorcodes Fixed the remaining help-override blocker in 4bc88efa3. The safe variant now recognizes help only before the first `--` end-of-options marker, so `codex-migrate export ... --apply -- --help` remains controlled by `command.codex-migrate.apply`. Added both ownership and enabled-policy regressions.

Validation:
- `80 passed` focused Codex Migrate suite
- `199 passed` extension/registry/directory/contribution/trust/policy suites
- ruff check + format check
- extension directory render/export freshness checks
- wheel build and packaged runtime/contribution inspection
- `git diff --check`

Please re-review when convenient.

**Review by kantorcodes — COMMENTED** [review]
`main` moved again. Please merge current `main` (`b910d77`) and regenerate the derived catalog/baseline/projection artifacts before the next pass; this repo requires branches to be current before merge.
